docs/false_positives.md

```diff
@@ -0,0 +1,244 @@
+* Prefer standard functions to custom solutions
+* Do not turn off `core` checks
+
+### <a name="correlated-conditions"></a> Correlated conditions
+
+The analyzer might not be able to detect that two conditions are mutually 
+exclusive. Let us consider the following false positive:
+
+```cpp
+int x = f();
+if (condition A) {
+  x = 0;
+}
+...
+if (nontrivial condition) {
+  g(5/x); // Division by zero warning.
+}
+```
+
+It can be rewritten to as the following to suppress the warning:
+
+```cpp
+int x = f();
+if (condition A) {
+  x = 0;
+}
+...
+if (nontrivial condition) {
+  assert(!(condition A));
+  g(5/x); // No warning.
+}
+```
+
+### <a name="partial-functions"></a> Partial functions
+
+Some functions only work on a set of possible input values. This precondition
+is unknown to the analyzer unless it is expressed explicitly in the code.
+
+Let us consider the following false positive:
+
+```cpp
+int f(MyEnum Val) {
+  int x = 0;
+  switch (Val) {
+    case MyEnumA: x = 1; break;
+    case MyEnumB: x = 5; break;
+  }
+  return 5/x; // Division by zero when Val == MyEnumC.
+}
+```
+
+It can be rewritten to as the following to suppress the warning:
+
+```cpp
+int f(MyEnum Val) {
+  int x = 0;
+  switch (Val) {
+    case MyEnumA: x = 1; break;
+    case MyEnumB: x = 5; break;
+    default: assert(false); break;
+  }
+  return 5/x; // No warning.
+}
+```
+
+Other macros or builtins expressing unreachable code may be used.
+Note that the rewritten code is also safer, since debug builds now check for
+more precondition violations. 
+
+### <a name="prefer-standard-functions"></a> Prefer standard functions
+
+The analyzer models the behavior of some standard functions but it has
+no knowledge about the semantics of custom made functions declared in a
+library that compiled separately or not reached by the analyzer.
+
+Let us consider the following false positive:
+
+```cpp
+int f() {
+  const char *Str = "MyStr";
+  int Val = 0;
+  ...
+  if (mystrlen(Str)) {
+    Val = 1;
+  }
+  return 5 / Val; // Division by zero.
+}
+```
+
+It can be rewritten to as the following to suppress the warning:
+
+```cpp
+int f() {
+  const char *Str = "MyStr";
+  int Val = 0;
+  ...
+  if (strlen(Str)) {
+    Val = 1;
+  }
+  return 5 / Val; // No warning.
+}
+```
+
+The result is easier to read for other developers who might not be familiar
+with the custom version of the function. The standard library functions also
+tend to be faster and more correct than custom solutions.
+
+### <a name="do-not-turn-off-core-checks"></a> Do not turn off `core` checks
+
+Checks in the static analyzer are usually not just reporting issues but
+also help modelling language constructs and library functions. You should never
+turn off checks from the `core` package.
+
+### <a name="suppress-specific-dead-store-warnings"></a> Suppress specific dead store warnings
+
+How to suppress a specific dead store warning and more useful tips can be found
```

Comment:
Suppressing dead stores.... in the Clang pilot at BBI in Kista I saw many false positives for such code:

    #if X1>10  // some condition
    #define X1_PRINT(X)   printf("%i",X);  // use X somehow
    #else
    #define X1_PRINT(X)  // Do nothing
    #endif
    
    void foo() {
        int x = 123;
        X1_PRINT(x);
    }

We could have worked on tweking the defines but we wanted to use a standard build configuration... Our solution was to create a define (I think __clang_analyzer__) and make sure the variable was used...

    #if X1>10  // some condition
    #define X1_PRINT(X)   printf("%i",X)  // use X somehow
    #else
    #ifdef __clang_analyzer__
    #define X1_PRINT(X)  (void)x
    #else
    #define X1_PRINT(X)  // Do nothing
    #endif
    #endif
    
    void foo() {
        int x = 123;
        X1_PRINT(x);
    }

Reply:
Maybe obvious... but here is another fix we did:

    void foo() {
        int x = 3;
    #ifdef ABC
        dostuff(x);
    #endif
    }

=>

    void foo() {
    #ifdef ABC
        int x = 3;
        dostuff(x);
    #endif
    }

I especially remember a case where we could fix the false positive by cleaning up the code.